### ISSUE TITLE
Reworking JavaScript documentation & TypeScript declarations

### ISSUE THREAD

**Issue by blutorange** [primefaces/primefaces#12500]
We document the JavaScript code with doc comments, then generate TypeScript declarations from that, and finally use the declarations to generate an API documentation.

That works well, but the current setup is sub-optimal and needs to be improved.

Why?

* We currently use a custom script to extract TypeScript declarations from the JavaScript works -- but it's complex. It also contains a non-trivial amount of custom, non-standard JSDoc syntax.
* Take a look at the readme and the source code here: https://github.com/primefaces/primefaces/tree/master/primefaces/src/main/type-definitions
* It uses TypeScript in a way that may break when TypeScript changes their API, and that may be hard to fix. See also #11539. I tried
  updating to the latest TypeScript version, which proved to be harder than expected due to TypeScript API changes.

The custom script was necessary to work with the existing JavaScript code structure. I evaluated some tools to generated TypeScript declarations from the existing JavaScript code structure, but none were up to the task. But now it's 2024, IE support is gone, so I think we should take a look at better approaches.

While I think we can mostly keep everything backwards-compatible, there are bound to be a few edge cases. So it should only be done as part of a new major version.

I haven't worked out all details yet, but the general approach would be to completely remove the custom documentation script; make some adjustment to the JavaScript files and then use TypeScript to generate the type declaration files. Finally, use the declaration files to generate the documentation, the same we are doing it now.

What do you think about it?

See below for details. **My two main questions:**

* Are we OK with shipping JavaScript code that contains some ES6 features, such as e.g. classes? Or do we need to support browsers that don't support these (are there even any such browsers anymore)?
* Should we keep the code in the PrimeFaces repo or use NPM to fetch the dependencies?

## Fix the JavaScript code structure

The current structure of the JavaScript code is hard to understand for tools. In particular:

* Using functions as classes and mutating prototypes
  * In fact, we use a custom class "polyfill" called [Simple JavaScript Inheritance by By John Resig](https://github.com/primefaces/primefaces/blob/2e7aad32e0a01de041976ecdec0ae907f2228ef4/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.widget.js#L6-L63) That's even harder to analyze.
* Assigning to the global scope everywhere. Most tools understand modules better than a bunch of global modifying script files. 

The first point is the most important one. All widgets are intended to be classes, but it's hard for tools like TypeScript to infer the structure of the intended class from functions and prototypes and the custom class polyfill we currently use.

Instead, we should just use a class directly, JavaScript supports those for almost 10 years now. That raises my first question:

* **Question** Are we OK with shipping JavaScript code that contains ES6 classes? If not, we need to add a production build step that downlevels JavaScript back to function prototypes. There are tools like [esbuild](https://esbuild.github.io/) that can do this extremly quickly; and the PrimeFaces resources optimizer plugin uses closure compiler (I believe), which should also be able to do this.

Secondly, I would use ES modules (`import` / `export`) to make dependencies explicit. I'll have to experiment with that a bit, especially how it interacts with quick dev builds and starting a dev server. For development, all browsers support ES modules natively, but the Jakarta Faces resource servlet mechanism does not play nicely with ES modules.

For example, [forms.inputtext.js](https://github.com/primefaces/primefaces/blob/2e7aad32e0a01de041976ecdec0ae907f2228ef4/primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.inputtext.js) would be changed to something like this. Note that the actual code remains the same -- only the structure is changed (the code should be changed if we want TypeScript to understand it for type checking, but as I said, that can be done later, if we want):

```js
// forms.inputtext.js

// @ts-nocheck

/** @import {PartialWidgetCfg} from "../core/core.widget.js"; */
/** @import {InputTextCfg} from "./forms.inputtext.types.js"; */

import * as PrimeFaces from "../core/core.js"
import * as utils from "../core/core.utils.js"
import { BaseWidget } from "../core/core.widget.js"

/**
 * __PrimeFaces InputText Widget__
 *
 * InputText is an extension to standard inputText with skinning capabilities.
 *
 * @template {InputTextCfg} TCfg
 */
export class InputText extends BaseWidget {
  /**
   * @param {PartialWidgetCfg<TCfg>} cfg
   */
  constructor(cfg) {
    super(cfg);
    PrimeFaces.skinInput(this.jq);

    /**
     * The DOM element for the counter that informs the user about the number of
     * characters they can still enter before they reach the limit.
     * @type {JQuery}
     */
    this.counter;

    //Counter
    if (this.cfg.counter) {
      this.counter = this.cfg.counter ? $(PrimeFaces.escapeClientId(this.cfg.counter)) : null;
      this.cfg.counterTemplate = this.cfg.counterTemplate || '{0}';
      this.updateCounter();

      if (this.counter) {
        var $this = this;
        this.jq.on('input.inputtext-counter', function (e) {
          $this.updateCounter();
        });
      }
    }
  }

  /**
   * Disables this input so that the user cannot enter a value anymore.
   */
  disable() {
    utils.disableInputWidget(this.jq);
  }

  /**
   * Enables this input so that the user can enter a value.
   */
  enable() {
    utils.enableInputWidget(this.jq);
  }

  /**
   * Updates the counter value that keeps count of how many more characters the user can enter before they reach the
   * limit.
   * @private
   */
  updateCounter() {
    var value = this.normalizeNewlines(this.jq.val()),
      length = this.cfg.countBytesAsChars ? PrimeFaces.utils.countBytes(value) : value.length;

    if (this.counter && this.cfg.maxlength) {
      var remaining = this.cfg.maxlength - length;
      if (remaining < 0) {
        remaining = 0;
      }

      var counterText = this.cfg.counterTemplate
        .replace('{0}', remaining)
        .replace('{1}', length)
        .replace('{2}', this.cfg.maxlength);

      this.counter.text(counterText);
    }
  }

  /**
   * Replaces all line breaks with a Window-style line break (carriage return + line feed).
   * @private
   * @param {string} text Text to normalize.
   * @return {string} The given text, with all line breaks replaced with carriage return + line feed.
   */
  normalizeNewlines(text) {
    return text.replace(/(\r\n|\r|\n)/g, '\r\n');
  }
}
```

In addition, we'd have a separate `*.d.ts` file with types need by the widget. It does not have any effect on the runtime. Using a separate file
gets rid of [having to pollute the header with long doc comments.](https://github.com/primefaces/primefaces/blob/2e7aad32e0a01de041976ecdec0ae907f2228ef4/primefaces/src/main/resources/META-INF/resources/primefaces/dialog/dialog.js#L7-L88)

```ts
// forms.inputtext.types.d.ts

/**
 * The configuration for the {@link  InputText| InputText widget}.
 * You can access this configuration via
 * {@link PrimeFaces.widget.BaseWidget.cfg|BaseWidget.cfg}. Please note that this
 * configuration is usually meant to be read-only and should not be modified.
 */
export interface InputTextCfg {
    /**
     * ID of the label component to display remaining and entered characters.
     */
    counter: string;
    /**
     * Template text to display in counter, default value is `{0}`.
     */
    counterTemplate: string;
}
```

## Type annotations via JSDocs

This is not so much a change, but rather a by-product we'd get for free.

TypeScript allows two ways how JavaScript code can be annotated with types and be type checked:

* The usual TypeScript *.ts files, which extend the grammar / syntax of Javascript. Type annotations are added inline after a colon `:`.
* [Type annotation](https://www.puruvj.dev/blog/get-to-know-typescript--using-typescript-without-typescript) [via JSDocs](https://alexharri.com/blog/jsdoc-as-an-alternative-typescript-syntax). This is basically what we already have -- types inside doc comments, e.g. see [forms.button.js](https://github.com/primefaces/primefaces/blob/2e7aad32e0a01de041976ecdec0ae907f2228ef4/primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.button.js#L18)

Recently, [there were some](https://www.codestackr.com/blog/typescript-vs-jsdoc-a-shift-in-the-development-landscape) major project that switched from TypeScript files back to plain JavaScript files, but are still type-checked via type annotation in JSDocs (some news articles called that "removing TypeScript", but [that's misleading at best](https://www.reddit.com/r/sveltejs/comments/12cyady/are_you_leaving_svelte_cause_they_are_ditching/)).  A major example would be [Svelte](https://github.com/sveltejs/svelte/pull/8569). One major [advantage](https://github.com/sveltejs/kit/discussions/4429#discussioncomment-2423814) of this approach is that it removes the required compile step from TypeScript to JavaScript. I also experimented with this at work and it works pretty well.

In theory, this means that we'd bascially just have to set a few flags when running TypeScript (to generate the type declarations for the documentation); and would have fully type checked code.

In practice, I'm certain there will be many type errors. Fortunately, you can just add `// @ts-nocheck` to the top of a file, which tells TypeScript not to fail on type errors in that file. This way, we can add type checking on a gradual basis, e.g. when editing a file anyways.

See also TypeScript's [JSDoc reference](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#constructor). One major drawback was that
importing types from other files was quite painful to write, but this was also recently remediated by support [for @import via a doc comment.](https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#the-jsdoc-import-tag)

A quick example how it looks like (again, note that this is what we already have!)

```js
/** @import {Integer} from "./helper-types" */

/**
 * @param {Integer} x 
 * @param {Integer} y 
 * @returns {Integer}
 */
export function sum(x, y) {
    return x + y;
}

/**
 * @param {Integer} x 
 * @param {Integer} y 
 * @returns {Integer}
 */
export function multiply(x,y ) {
    return x* y;
}
```

## Reconsider whether we want to inline 3rd party code

This is optional, but I though I'd bring it up and ask for your opinion.

Currently, all 3rd party code is added directly to the PrimeFaces repo, see e.g. [signature](https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/resources/META-INF/resources/primefaces/signature/0-signature.js).

We can keep it this way, but the alternative would be to use NPM, i.e. just declare the needed 3rd party libraries in a `package.json`. For development, there are several options such as building a "vendor bundle" that contains all code from these 3rd party libraries. You would only have to build it once, to speed up development.

* **Question** What do you think, should we keep the code in the PrimeFaces repo or use NPM to fetch the dependencies?

**Comment by melloware** [comment]
There is a lot here @blutorange but a few notes/thoughts.

1. I don't think we can use `npm` and bring things in like Signature because JQ Signature and JQ FileUpload are no longer receiving updates so to get it working with JQuery 4.0.0 we had to patch things here manually in our code.  I suspect the same will go for more and more of our plugins where our versions are the gold copy to work with PF.  So their NPM versions are way out of date.

2. As far a John Resig `polyfill` for classes i would love to get rid of this.  Maybe this is a small baby step we can take in the right direction as a standalone issue?  Replace this polyfill with JS classes?

3. We only support ES6 browsers now so using ES6 features is perfectly acceptable.

4. I worry about using NPM and using imports.  Not sure how this impacts "monkeypatching" which a lot of users have in their production systems to make PF behave the way they want it i wouldn't want to break backward compatibility.  Not sure if that is possible with what you are proposing?

**Comment by blutorange** [comment]
Thanks for the feedback (and for reading everything).

> We only support ES6 browsers now so using ES6 features is perfectly acceptable.

That's awesome! : )

> I worry about using NPM and using imports. Not sure how this impacts "monkeypatching" which a lot of users have in their production systems to make PF behave the way they want it i wouldn't want to break backward compatibility. Not sure if that is possible with what you are proposing?

For the final JavaScript that gets included in the JAR file, it should definitely be bundled into a single file, removing imports. Everything will still be available via the global scope, e.g. `PrimeFaces.widget.SelectOneMenu` etc. Everything else would to too breaking of a change. So everything should stay compatible.

Modules and imports help greatly with code structure, though, and to get tools to understand the code structure. What I meant was: For development (not production), browsers support modules, so no additional build step would be needed for development.

> As far a John Resig polyfill for classes i would love to get rid of this. Maybe this is a small baby step we can take in the right direction as a standalone issue? Replace this polyfill with JS classes?

That's a good idea. But the problem is, we'd then have to adjust the documentation script to work with classes (which is useless work, as we want to get rid of it). But perhaps we could do the following:

* (separate branch) Proof of concept. Update only 2-3 widgets + core script, and setup the build to generate type declarations + documentation.
* (main branch) Once we know the approach actually works, remove the old documentation script in the main branch and update to the new class syntax
* (separate branch) Finish what's left for the type declarations + documentation.

> I don't think we can use npm and bring things in like Signature because JQ Signature and JQ FileUpload are no longer receiving updates so to get it working with JQuery 4.0.0 we had to patch things here manually in our code. I suspect the same will go for more and more of our plugins where our versions are the gold copy to work with PF. So their NPM versions are way out of date.

Hmm, I see. But in general, you would be fine with NPM if possible?

I see some minified 3rd party lib which I suppose nobody modified. We could pull those from NPM; and only include libs in the PrimeFaces repo that were actually modified. Would also help letting people know which files had to be modified (and where we are just using the default version from NPM). 

As for the the need to fix third party libraries, some (theoretical) options that come to my mind: 

* When the amount of changes is small: We can use [NPM patch package](patch-package) or [yarn patch](https://yarnpkg.com/cli/patch). Essentially, we just add a `*.patch` file to our repo that includes the changes.
  * The main benefit is that it's immediately obvious what was changed compared to the version in the NPM repos.
  * Obviously that's not practical when the lib needs to be modified substantially.
* Make a fork of the lib, publish that to npm, then use that.
  * I suppose we don't want that, though, since it's too much overhead and additional work.

**Comment by melloware** [comment]
I will let some others chime in here.

**Comment by blutorange** [comment]
If there aren't any objections, I'll go ahead and make a proof-of-concept, then we can discuss further based on that.

**Comment by melloware** [comment]
That sounds good to me.

**Comment by blutorange** [comment]
Well, I started by making a separate JavaScript test project to test out ideas how the JavaScript and folder layout should / could look like.

https://github.com/blutorange/primefaces-js-refactor-test

In short:

* A separate folder for the core JavaScript and one folder for the additional widgets
* Widgets as proper classes, additional `*.d.ts` files for additional types
* The files all use `import` and `export` to reference each other.
* There's a `packages/core/closure.sh` and `packages/widgets/closure.sh` that runs closure compiler on the JavaScript files
  * Once it's done, it produces exactly the same files as now, i.e. a `core.js`, `components.js` + some separate files for some widgets such as `datepicker.js` and `signature.js`.
* We can just run [typedoc](https://typedoc.org/) on the files to produce documenation without custom scripts that parse and analyse the code.
  * You can checkout how the documentation looks like here https://blutorange.github.io/primefaces-js-refactor-test/
* There a minor TypeScript bug I encountered (microsoft/TypeScript#60113) where it doesn't output type parameters for `extends` clauses, but I hope that gets fixed soon.

**Comment by melloware** [comment]
the docs look really nice.   Here is a question for simpler widgets instead of `*.d.ts` couldn't they just be native `.ts` files?  For example `TagCloud` is a pretty small widget?

**Comment by blutorange** [comment]
>  the docs look really nice. Here is a question for simpler widgets instead of *.d.ts couldn't they just be native .ts files? For example TagCloud is a pretty small widget?

I'm not quite sure what you mean, but:

The difference between `.d.ts` and `.ts` files is that `.d.ts` can only contain types; and `.ts` files can contain actual code (functions etc.). Then people could accidentally write code that won't ever be run.

On the other hand: If we decide we are OK with writing TypeScript, we could combine the `.js` and `.d.ts` files into a single `.ts` file. That would simplify the file layout a bit. But then we'd have to use something like [esbuild](https://esbuild.github.io/), since closure compile can't process TS files. (Which might not be a bad idea, esbuild is faster than closure compiler and some recent benchmarks such as [this one](https://github.com/privatenumber/minification-benchmarks?tab=readme-ov-file) would suggest the difference in file size is neglibile).

**Comment by melloware** [comment]
Yes i mean convert the widget from .js to .ts so you don't need the separate `.d.ts` file it would just be 1 typescript file?   We don't have to do ALL the widgets but maybe we start with simpler ones and doing them in straight TypeScript?

**Comment by blutorange** [comment]
Ok, I see. Using TypeScript directly would make some things easier. I guess I'll convert the core, but keep some of the longer widget as JavaScript (I'm looking at you, data table ; )   I'll take another at the demo project to see how it would look with `ts` files.

And to reduce merge conflicts, I like the idea of starting with the conversion to classes first and commit that to main.

**Comment by melloware** [comment]
Yep of course.  I was just trying to prevent having a .js and a .d.ts to simpler widgets etc. Save editing two files when making chnages.  However datatable.js is a beast so obviously not that one!
